File: Ai App/voice/voice_controller.py
"""
Voice Controller module for Wolf AI Assistant
Manages the integration of speech recognition, text-to-speech, and wake word detection.
"""

import logging
import threading
import queue
import json
from typing import Optional, Callable, Dict, Any

class VoiceController:
    """Main controller for voice interaction functionality."""
    
    def __init__(self, 
                 wake_word: str = "hey wolf",
                 speech_timeout: int = 5,
                 speech_phrase_limit: int = 10,
                 tts_rate: int = 200,
                 tts_volume: float = 1.0,
                 tts_voice: str = None):
        """Initialize the voice controller.
        
        Args:
            wake_word: The wake word/phrase to use
            speech_timeout: Timeout in seconds for speech recognition
            speech_phrase_limit: Maximum length of a speech phrase in seconds
            tts_rate: Speech rate for text-to-speech
            tts_volume: Volume level for text-to-speech (0.0 to 1.0)
            tts_voice: Voice ID to use for text-to-speech
        """
        self.wake_word = wake_word
        self.speech_timeout = speech_timeout
        self.speech_phrase_limit = speech_phrase_limit
        self.logger = logging.getLogger(__name__)
        
        # Initialize components
        self.speech_to_text = None
        self.text_to_speech = None
        self.wake_word_detector = None
        
        # State
        self.is_listening = False
        self.is_processing = False
        self.callbacks = {
            'on_wake': [],
            'on_speech': [],
            'on_tts_start': [],
            'on_tts_end': [],
            'on_error': []
        }
        
        # Initialize components
        self._initialize_components(tts_rate, tts_volume, tts_voice)
        
        self.logger.info("VoiceController initialized")
    
    def _initialize_components(self, tts_rate: int, tts_volume: float, tts_voice: str):
        """Initialize the voice components."""
        try:
            from .speech_to_text import SpeechToText
            from .text_to_speech import TextToSpeech
            from .wake_word import WakeWordDetector
            
            self.speech_to_text = SpeechToText()
            self.text_to_speech = TextToSpeech(rate=tts_rate, volume=tts_volume, voice=tts_voice)
            self.wake_word_detector = WakeWordDetector(wake_word=self.wake_word)
            
            # Set up wake word detection callback
            self.wake_word_detector.callback = self._on_wake_word_detected
            
        except Exception as e:
            self.logger.error(f"Failed to initialize voice components: {e}", exc_info=True)
            raise
    
    def start(self):
        """Start the voice controller."""
        if self.is_listening:
            self.logger.warning("Voice controller is already running")
            return
            
        try:
            # Start wake word detection
            self.wake_word_detector.start()
            self.is_listening = True
            self.logger.info("Voice controller started")
            
        except Exception as e:
            self.logger.error(f"Failed to start voice controller: {e}", exc_info=True)
            self._trigger_callbacks('on_error', {'error': str(e)})
    
    def stop(self):
        """Stop the voice controller."""
        if not self.is_listening:
            return
            
        try:
            self.wake_word_detector.stop()
            self.text_to_speech.stop()
            self.is_listening = False
            self.is_processing = False
            self.logger.info("Voice controller stopped")
            
        except Exception as e:
            self.logger.error(f"Error stopping voice controller: {e}", exc_info=True)
    
    def toggle_listening(self, force_state: Optional[bool] = None) -> bool:
        """Toggle listening state.
        
        Args:
            force_state: If provided, force the listening state to this value
            
        Returns:
            bool: The new listening state
        """
        new_state = not self.is_listening if force_state is None else force_state
        
        if new_state:
            self.start()
        else:
            self.stop()
            
        return self.is_listening
    
    def speak(self, text: str, callback: Optional[Callable[[], None]] = None):
        """Speak the given text.
        
        Args:
            text: The text to speak
            callback: Optional function to call when speaking is done
        """
        self._trigger_callbacks('on_tts_start', {'text': text})
        
        def on_speak_done():
            self._trigger_callbacks('on_tts_end', {'text': text})
            if callback:
                callback()
        
        self.text_to_speech.speak(text, on_speak_done)
    
    def process_text_command(self, text: str):
        """Process a text command.
        
        Args:
            text: The text command to process
        """
        self._trigger_callbacks('on_speech', {'text': text})
    
    def _on_wake_word_detected(self):
        """Handle wake word detection."""
        if self.is_processing:
            return
            
        self._trigger_callbacks('on_wake', {})
        self._start_speech_recognition()
    
    def _start_speech_recognition(self):
        """Start listening for a voice command."""
        if self.is_processing:
            return
            
        self.is_processing = True
        self.logger.info("Listening for voice command...")
        
        def recognition_callback(text: str):
            if text:
                self.process_text_command(text)
            self.is_processing = False
            
        self.speech_to_text.listen_in_background(
            recognition_callback,
            timeout=self.speech_timeout,
            phrase_time_limit=self.speech_phrase_limit
        )
    
    def register_callback(self, event: str, callback: Callable[[Dict[str, Any]], None]):
        """Register a callback for voice events.
        
        Args:
            event: The event to listen for ('on_wake', 'on_speech', 'on_tts_start', 'on_tts_end', 'on_error')
            callback: Function to call when the event occurs
        """
        if event in self.callbacks:
            self.callbacks[event].append(callback)
        else:
            self.logger.warning(f"Unknown event type: {event}")
    
    def _trigger_callbacks(self, event: str, data: Dict[str, Any]):
        """Trigger all callbacks for an event.
        
        Args:
            event: The event that occurred
            data: Data to pass to the callbacks
        """
        for callback in self.callbacks.get(event, []):
            try:
                callback(data)
            except Exception as e:
                self.logger.error(f"Error in {event} callback: {e}", exc_info=True)
    
    def update_settings(self, settings: Dict[str, Any]):
        """Update voice settings.
        
        Args:
            settings: Dictionary of settings to update
        """
        if 'wake_word' in settings:
            self.wake_word = settings['wake_word']
            self.wake_word_detector.update_wake_word(self.wake_word)
            
        if 'tts_rate' in settings:
            self.text_to_speech.set_rate(settings['tts_rate'])
            
        if 'tts_volume' in settings:
            self.text_to_speech.set_volume(settings['tts_volume'])
            
        if 'tts_voice' in settings:
            self.text_to_speech.set_voice(settings['tts_voice'])
    
    def shutdown(self):
        """Shut down the voice controller and release resources."""
        self.stop()
        
        if hasattr(self, 'text_to_speech') and self.text_to_speech:
            self.text_to_speech.shutdown()
        
        self.logger.info("Voice controller shut down")
    
    def __del__(self):
        """Ensure resources are cleaned up."""
        self.shutdown()

File: Ai App/voice/voice_controller.js
/**
 * Voice Controller for Wolf AI Assistant (JavaScript Version)
 * Manages voice input/output and wake word detection
 */

class VoiceController {
    /**
     * Initialize the voice controller
     * @param {Object} options - Configuration options
     */
    constructor(options = {}) {
        // Default settings
        this.settings = {
            wakeWord: options.wakeWord || 'hey wolf',
            speechTimeout: options.speech_timeout || 5,
            speechPhraseLimit: options.speech_phrase_limit || 10,
            ttsRate: options.tts_rate || 200,
            ttsVolume: options.tts_volume || 1.0,
            ttsVoice: options.tts_voice || null
        };

        // State
        this.isListening = false;
        this.isInitialized = false;
        this.callbacks = {
            on_wake: [],
            on_speech: [],
            on_tts_start: [],
            on_tts_end: [],
            on_error: []
        };

        // WebSocket connection
        this.ws = null;
        this.wsUrl = 'ws://localhost:5001';

        // Initialize
        this.initialize();
    }

    /**
     * Initialize the voice controller
     */
    async initialize() {
        try {
            // Check for Web Speech API support
            if (!('webkitSpeechRecognition' in window) && !('SpeechRecognition' in window)) {
                throw new Error('Speech recognition not supported in this browser');
            }

            // Set up WebSocket connection
            this.setupWebSocket();

            // Set up speech recognition
            this.setupSpeechRecognition();

            // Set up text-to-speech
            this.setupTextToSpeech();

            this.isInitialized = true;
            console.log('Voice controller initialized');
        } catch (error) {
            console.error('Error initializing voice controller:', error);
            this.triggerCallback('on_error', { error: error.message });
        }
    }

    /**
     * Set up WebSocket connection to backend
     */
    setupWebSocket() {
        try {
            this.ws = new WebSocket(this.wsUrl);

            this.ws.onopen = () => {
                console.log('WebSocket connected to voice backend');
                // Send initialization message
                this.sendWebSocketMessage('init', {
                    wakeWord: this.settings.wakeWord,
                    speechTimeout: this.settings.speechTimeout,
                    speechPhraseLimit: this.settings.speechPhraseLimit
                });
            };

            this.ws.onmessage = (event) => {
                try {
                    const message = JSON.parse(event.data);
                    this.handleWebSocketMessage(message);
                } catch (error) {
                    console.error('Error processing WebSocket message:', error);
                }
            };

            this.ws.onclose = () => {
                console.log('WebSocket connection closed');
                // Try to reconnect after a delay
                setTimeout(() => this.setupWebSocket(), 3000);
            };

            this.ws.onerror = (error) => {
                console.error('WebSocket error:', error);
                this.triggerCallback('on_error', { error: 'WebSocket connection error' });
            };
        } catch (error) {
            console.error('Error setting up WebSocket:', error);
            this.triggerCallback('on_error', { error: 'Failed to connect to voice backend' });
        }
    }

    /**
     * Set up speech recognition
     */
    setupSpeechRecognition() {
        const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition;
        this.recognition = new SpeechRecognition();
        
        this.recognition.continuous = true;
        this.recognition.interimResults = false;
        this.recognition.lang = 'en-US';

        this.recognition.onresult = (event) => {
            const transcript = event.results[event.results.length - 1][0].transcript;
            if (transcript.trim().toLowerCase().includes(this.settings.wakeWord.toLowerCase())) {
                this.triggerCallback('on_wake');
            } else {
                this.triggerCallback('on_speech', { text: transcript });
            }
        };

        this.recognition.onerror = (event) => {
            console.error('Speech recognition error:', event.error);
            this.triggerCallback('on_error', { error: `Speech recognition error: ${event.error}` });
        };
    }

    /**
     * Set up text-to-speech
     */
    setupTextToSpeech() {
        if (!('speechSynthesis' in window)) {
            throw new Error('Text-to-speech not supported in this browser');
        }
        
        this.synth = window.speechSynthesis;
        this.utterance = new SpeechSynthesisUtterance();
        
        this.utterance.onstart = () => {
            this.triggerCallback('on_tts_start', { text: this.utterance.text });
        };
        
        this.utterance.onend = () => {
            this.triggerCallback('on_tts_end');
        };
        
        this.utterance.onerror = (event) => {
            console.error('TTS error:', event);
            this.triggerCallback('on_error', { error: 'Text-to-speech error' });
        };
    }

    /**
     * Toggle listening state
     */
    toggle_listening() {
        if (!this.isInitialized) {
            console.error('Voice controller not initialized');
            return false;
        }

        if (this.isListening) {
            this.stopListening();
        } else {
            this.startListening();
        }
        
        return this.isListening;
    }

    /**
     * Start listening for voice input
     */
    startListening() {
        if (this.isListening || !this.recognition) return;
        
        try {
            this.recognition.start();
            this.isListening = true;
            console.log('Started listening for voice input');
        } catch (error) {
            console.error('Error starting speech recognition:', error);
            this.triggerCallback('on_error', { error: 'Failed to start listening' });
        }
    }

    /**
     * Stop listening for voice input
     */
    stopListening() {
        if (!this.isListening || !this.recognition) return;
        
        try {
            this.recognition.stop();
            this.isListening = false;
            console.log('Stopped listening for voice input');
        } catch (error) {
            console.error('Error stopping speech recognition:', error);
        }
    }

    /**
     * Speak text using text-to-speech
     * @param {string} text - The text to speak
     */
    speak(text) {
        if (!this.synth || !this.utterance) {
            console.error('Text-to-speech not available');
            return;
        }
        
        try {
            // Cancel any ongoing speech
            this.synth.cancel();
            
            // Set up the utterance
            this.utterance.text = text;
            this.utterance.rate = this.settings.ttsRate / 100;
            this.utterance.volume = this.settings.ttsVolume;
            
            // Set voice if available
            if (this.settings.ttsVoice) {
                const voices = this.synth.getVoices();
                const voice = voices.find(v => v.name === this.settings.ttsVoice);
                if (voice) {
                    this.utterance.voice = voice;
                }
            }
            
            // Speak
            this.synth.speak(this.utterance);
        } catch (error) {
            console.error('Error in text-to-speech:', error);
            this.triggerCallback('on_error', { error: 'Failed to speak text' });
        }
    }

    /**
     * Register a callback for voice events
     * @param {string} event - Event name ('on_wake', 'on_speech', 'on_tts_start', 'on_tts_end', 'on_error')
     * @param {Function} callback - Callback function
     */
    register_callback(event, callback) {
        if (this.callbacks[event]) {
            this.callbacks[event].push(callback);
        } else {
            console.warn(`Unknown event type: ${event}`);
        }
    }

    /**
     * Trigger callbacks for an event
     * @private
     */
    triggerCallback(event, data = {}) {
        if (!this.callbacks[event]) return;
        
        this.callbacks[event].forEach(callback => {
            try {
                callback(data);
            } catch (error) {
                console.error(`Error in ${event} callback:`, error);
            }
        });
    }

    /**
     * Send a message via WebSocket
     * @private
     */
    sendWebSocketMessage(type, data = {}) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify({ type, ...data }));
        } else {
            console.warn('WebSocket not connected');
        }
    }

    /**
     * Handle incoming WebSocket messages
     * @private
     */
    handleWebSocketMessage(message) {
        if (!message || !message.type) return;

        switch (message.type) {
            case 'wake_word_detected':
                this.triggerCallback('on_wake');
                break;
            case 'speech_recognized':
                this.triggerCallback('on_speech', { text: message.text });
                break;
            case 'tts_start':
                this.triggerCallback('on_tts_start', { text: message.text });
                break;
            case 'tts_end':
                this.triggerCallback('on_tts_end');
                break;
            case 'error':
                this.triggerCallback('on_error', { error: message.error });
                break;
            default:
                console.warn('Unknown message type:', message.type);
        }
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = VoiceController;
}
